presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,775 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Range;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.SliceInput;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.Immutable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.presto.execution.buffer.BufferResult.emptyResults;
+import static com.facebook.presto.execution.buffer.BufferState.FINISHED;
+import static com.facebook.presto.execution.buffer.BufferState.FLUSHING;
+import static com.facebook.presto.execution.buffer.BufferState.NO_MORE_BUFFERS;
+import static com.facebook.presto.execution.buffer.BufferState.OPEN;
+import static com.facebook.presto.execution.buffer.OutputBuffers.BufferType.SPOOLING;
+import static com.facebook.presto.spi.StandardErrorCode.SPOOLING_STORAGE_ERROR;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readSerializedPages;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterators.advance;
+import static com.google.common.collect.Range.closedOpen;
+import static com.google.common.util.concurrent.Futures.catchingAsync;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.Futures.transform;
+import static com.google.common.util.concurrent.Futures.transformAsync;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+@SuppressWarnings("UnstableApiUsage")
+public class SpoolingOutputBuffer
+        implements OutputBuffer
+{
+    private final TaskId taskId;
+    private final String taskInstanceId;
+    private final OutputBuffers outputBuffers;
+    private final StateMachine<BufferState> state;
+    private final TempDataOperationContext tempDataOperationContext;
+    private final TempStorage tempStorage;
+    private final DataSize threshold;
+    private final FinalizerService finalizerService;
+    private final ListeningExecutorService executor;
+
+    private final AtomicLong totalBufferedBytes = new AtomicLong();
```

Comment:
All these variables are used in synchronized blocks. I am not sure why they are atomic. Option 1: we can reduce the synchronized and leverage from the atomicity. Option 2: we can use primitives.